### DISCUSSION TITLE
Unexpected behavioral difference between selectinload and joinedload with pending mutations

### DISCUSSION THREAD

**Discussion by vineetgopal** [sqlalchemy/sqlalchemy#12058] [Usage Questions] [answered]
Hello! I recently noticed a behavioral difference between how selectinload and joinedload work with pending mutations, and I'm wondering if this difference is intentional or a bug.

**Overall context**
We use joinedload as a mechanism for loading a relationship on multiple models that are already in the session, with a single query. However, I noticed that these relationships were not getting populated as I expected when there were pending mutations to the relationship.

```
# This query loads the `children` relationship on all Parent models, including those with pending changes.
session.query(Parent).options(selectinload(Parent.children)).all()

# This query _does not_ load the `children` relationship on any Parent models with pending changes on the children relationship
session.query(Parent).options(joinedload(Parent.children)).all() 
```

**Minimal reproduction**
We use SQLAlchemy 1.4.51, but I was able to reproduce this on SQLAlchemy versions 1.3.24, 1.4.51, and 2.0.36.

I included 4 tests: the first 3 pass, and the last one (joinedload with pending mutations) fails.

```
from sqlalchemy import Column, create_engine, ForeignKey, inspect, Integer
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import backref, joinedload, relationship, selectinload, Session

Base = declarative_base()

class Parent(Base):
    __tablename__ = "parent"
    id = Column(Integer, primary_key=True)

class Child(Base):
    __tablename__ = "child"
    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer, ForeignKey("parent.id"))
    parent = relationship("Parent", backref=backref("children"))

engine = create_engine(f"sqlite:///:memory:", echo="debug")
Base.metadata.create_all(engine)

session = Session(engine, autoflush=False)
session.commit()

parent = Parent()
child = Child(parent=parent)
session.add_all([parent, child])
session.commit()

# Test #1: Load the relationship directly with pending changes
session.expunge_all()
parent = session.query(Parent).one()
session.add(Child(parent=parent))
assert "children" not in inspect(parent).dict  # Relationship is not loaded
parent.children
assert "children" in inspect(parent).dict  # Relationship is loaded

# Test #2: selectinload with pending changes
session.expunge_all()
parent = session.query(Parent).one()
session.add(Child(parent=parent))
assert "children" not in inspect(parent).dict  # Relationship is not loaded
session.query(Parent).options(selectinload(Parent.children)).all()
assert "children" in inspect(parent).dict  # Relationship is loaded

# Test #3: joinedload without pending changes
session.rollback()
session.expunge_all()
parent = session.query(Parent).one()
assert "children" not in inspect(parent).dict  # Relationship is not loaded
session.query(Parent).options(joinedload(Parent.children)).all()
assert "children" in inspect(parent).dict  # Relationship is loaded

# Test #4: joinedload with pending changes
session.rollback()
session.expunge_all()
parent = session.query(Parent).one()
session.add(Child(parent=parent))
assert "children" not in inspect(parent).dict  # Relationship is not loaded
session.query(Parent).options(joinedload(Parent.children)).all()
assert (
    "children" in inspect(parent).dict
)  # This assertion fails. Accessing `parent.children` will do another DB query
```

**Comment by zzzeek** [comment, marked as answer]
this is an implementation artifact, the mechanism by which "selectinload" works is called a "post load", that is, a step within the loading process that emits a second query.    in this case it would be firing off for this object because the "children" collection was otherwise not loaded.   it has the ability to merge the so-called "pending backrefs", that is, your pending `Child` object, into the collection that it creates, so that you get the right answer at the end, a list with two `Child` objects where one is still pending (unflushed).

"joinedload" takes a different codepath where it will only populate into a collection that's fully unloaded.   the presence of pending "Child" considers the collection to be already present, even though it's not present in the object's dictionary (it's in a collection on the object's `InstanceState` called "pending mutations").   joinedload does not make use of the same utility that selectinload does in order to populate collections and it does not include the behavior of loading joined rows while also reconciling pending mutations.

so that's why peeking at the internal implementation of the collection, the timing and method by which results are loaded is different.

**Reply by vineetgopal** [reply]
Thanks @zzzeek for explaining the behavior. That makes sense, and is aligned with the behavior I see.

Since I already have the parent tables loaded in, I was hoping to leverage native SQLAlchemy functionality to bulk-load the children with a single query:
```
session.query(Parent).filter(Parent.id.in_([parent.id for parent in parents]))
.options(load_only(Parent.id).lazyload("*"), joinedload(Parent.children))
.all()
```

This is obviously not that important in my minimal example, but helpful when there are lots of relationships and larger tables. I can use `selectinload` here, but as you mentioned, that will do 2 queries.

I did find [this recipe](https://github.com/sqlalchemy/sqlalchemy/wiki/DisjointEagerLoading) from [this discussion](https://github.com/sqlalchemy/sqlalchemy/discussions/11901), though am a bit wary of implementing it since it'd require ensuring we handle each relationship the same way that SQLAlchemy does (order, etc.).

I'll probably just go with `selectinload` since it seems safer, but do you know if it's possible to achieve the single-query goal in any other way?

**Reply by zzzeek** [reply]
you can always add [populate_existing](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#orm-queryguide-populate-existing) and everything that can load will load, but it would erase your pending change if you didnt flush it first

**Reply by vineetgopal** [reply]
Ah, thanks for the pointer. I was looking for as close of a replacement for:
```
for parent in parents:
    parent.children
```

but in just one query instead of N queries. It looks like `selectinload` is the closest to that (handles pending changes, etc.), so I'll stick with that.
